### ISSUE TITLE
New shiny Imgproc module in OpenCV 5

### ISSUE THREAD

**Issue by vpisarev** [opencv/opencv#25012]
### Describe the feature and motivation

# The new shiny Imgproc module in OpenCV 5.0

Please, look at #25011 first, as the ideas for Imgproc improvements have the same rationale and share many parts with the Core module.

Also, look at #25001, where it's suggested to remove some obsolete functionality.

So, the goals for the ultimate Imgproc module are mostly the same:

* provide essential set of image processing functions (of course, there will never be enough of such functions)
* highly efficient (taking advantage of the modern vector CPUs and GPUs) with small overhead
* compact size and small footprint

In the era of deep learning the classical image processing may still be useful, but for somewhat different purposes, not to solve big complex problems (which are usually addressed by deep models nowadays). So, what are the possible uses of imgproc today:

* image augmentation to train robust models that can tolerate variations in lighting conditions, color response, noise, distorted geometry etc. See https://pytorch.org/vision/stable/transforms.html for example of such a framework. In opencv_contrib we also have a draft version of a similar framework: https://github.com/opencv/opencv_contrib/pull/3335.
* pre-/post-processing before/after running a deep model. In DNN module we already have a family of functions `blobFromImage(s)` that perform typical preprocessing for some typical computer vision models, but the pipeline can also potentially include other operations, such as demosaicing (converting Bayer pattern to RGB), denoising, extracting crop (not necessarily rectangular) from image after running detection model and before running specialized model on a ROI, finding contour of an object after running semantic segmentation model (e.g. Segment Anything) etc.
* simple processing or more complex processing of an image when the task is relatively simple (e.g. just adjust the brightness or color or sharpness) or when the processed image is huge and so running deep net would take a lot of time.
* visualization of the results e.g. when overlay masks, contours, bounding boxes, body skeleton segments and facial features etc. can be drawn on top of input image to illustrate the results of algorithm.

The current imgproc already covers most of the use cases, but certain things still can be improved.

### Common improvements

* Enable batch processing in most of the image processing functions. Batch processing lets user to take the most of current many-core CPUs and even more powerful GPUs. When processed images are rather small, e.g. 'just' 640x480 or less, it's often ineffective to use more than 2-4 CPU core or a tiny fraction of GPU. The rest of the devices usually sit idle unless users organize quite sophisticated multi-threaded asyncronous pipeline on top of it. With native embedded batch support the usage scheme can be trivial. Users just collect several frames from video or a batch for training and then apply a serie of transforms using normal sequential code.
* Enable parallel processing where possible. Of course, if we have a big batch, the threading scheme looks trivial. But it's usually not difficult to make the scheme adaptive. In the case of big batch each thread processes one or several images, while in the case of a single big image each thread processes one or several tiles. Usually for a group of functions (e.g. all filters in OpenCV) it's enough to have one utility function or a class to organize such a smart job partioning scheme.
* Make use of FP16/BF16 type, AVX2/AVX512 instructions on modern CPUs (use dispatching more actively). FP16 is useful as input/output type on platforms that support hardware conversion FP16<=>FP32 (such as Intel/AMD with AVX2/AVX512 or ARMv8). Also, it's useful as intermediate type for processing UINT8 images, which is the most common case in optimized pipelines. In the latter case ARMv8.2+FP16 is needed, other platforms have limited support for FP16 arithmetics, unfortunately.
* CPU HAL, in the form of primitive calls to IPP, Carotene, custom HAL and in the form of OpenCV's Scalable Universal Intrinsics is already actively used. For OpenCV 5 it's time to introduce a special non-CPU HAL (which should be extension of OpenCL's T-API to other frameworks, such as CUDA), **link TBD**, so that all image processing functions can run faster using various accelerators.

Now let's discuss various specific improvements in Imgproc

### Filters, Edge Detection

* Get rid of FilterEngine. For some filters, like bilateral filter, we already have a nice parallel implementation. But for many other filter we still use very complex and very slow sequential FilterEngine. It's time to get rid of it. All filters should be parallel.
* Add guided filters from opencv_contrib. This is very nice and very fast alternative to bilateral filters and not just bilateral filters.
* Sharpening filter (unsharp mask-based) would also be nice to add.
* Maybe Susan edge detector, which is superior to Canny edge detector, would be nice to add (finally): https://users.fmrib.ox.ac.uk/~steve/susan/susan/susan.html.
* Add support for FP16/BF16 to filters.

### Image Warping

* Currently, fixed-point 'remap' mapping is computed and stored to a temporary buffer and then remap kernel is called. For WarpAffine, WarpPerspective and FP remap we can avoid temporary buffers, at the cost of increasing code size, but it's worth it, since functionality is essential, especially for image augmentation.
* Add support for FP16/BF16.

### Resize

Resize is one of the most important functions in the whole OpenCV. We should offer the best and the fastest implementation of resize.

* Probably, need to add an interpolation mode, which is a mix between INTER_AREA and INTER_LINEAR. We use intermediate integer scale factor to bring image size down and then scale it to the final size using bilinear interpolation. That should give quality comparable to INTER_AREA (e.g. moire-free), but at faster speed, because general-case INTER_AREA is quite slow. Note that smart implementation can compute 1st stage and then the 2nd one on each tile, w/o explictly forming complete intermediate image.
* Multi-step image upscale algorithms are also welcome to explore. We should also accelerate Bicubic and Lancos interpolation for image upscaling
* Add dynamic dispatchers with AVX2, AVX512, NEON+FP16 modes at least.
* Develop new smart tests that could tolerate deviations in resize implementation. We really need to enable faster resize in IPP and other potential HAL accelerators. Now 3rd-party resize implementations are often disabled as they don't pass tests. There should be special flags to resize to disable 3rd-party implementations for some accuracy-critical functions.

### Color conversion

* Refactor `cvtColor`. Now the list of colorspace conversion codes is huge and yet the function does not cover some potentially useful cases. cvtColor should be converted into a automatically constructed pipeline of color-processing element-wise operations. User should specify input color space (sometimes it's not just RGB, but sRGB or maybe AdobeRGB or whatever), order of input channels (in the case of RGB or YUV), subsampling of input channels in the case of YUV (4:2:0 or 4:2:2 or 4:4:4) etc. and the same is for the output color space, including also potential scaling and type conversion. Maybe some extra transformations could be performed in between, e.g. color temperature shift or brightness/contrast adjustments or maybe a custom LUT etc. It all sounds complex, but in reality it will be just a sequence of pre-programmed kernels that are subsequently called on each tile. With element-wise operations it's not difficult to implement.
* It would be extremely useful to integrate support for color management into imgproc: https://github.com/opencv/opencv_contrib/blob/4.x/modules/mcc/include/opencv2/mcc/ccm.hpp. This will make this part of OpenCV truly professional.

### Bits of computational photography

There is a dedicated Photo module for that, but for preprocessing we may probably need some basic image enhancement algorithms in imgproc module (like already existing bilateralFilter):

* More general functions to adjust brightness/contrast/color etc., like generalized LUT (even 3D LUT) with various kind of interpolation would be useful to have.
* Some simple image denoising methods, including non-local means or DCT-based denoising.
* De-fogging filter (that increases crispness & contrast) is another useful preprocessing kernel. cv::equalizeHist is a very simple method to do it, we may need something a little more advanced.
* 1-2 automatic white balance algorithms would also be useful to have, e.g. gray world or white point or whatever (maybe based on face detector and skintone analysis). Maybe it would be useful to specify explicit color temperature as parameter of image demosaicing algorithm.

### Drawing

* OpenCV drawing functions are amonst the most popular ones. And yet they are very old and quite slow. There are much faster and more flexible methods to render 2D antialiased figures, patch will be provided soon.
* As requested in the comment at #25001, support for floating-point coordinates would be nice to add. Maybe coordinates in API should be converted to float's or even double's.
* Text rendering has been already substantially improved in OpenCV 5.0-pre with added support for Truetype fonts (https://github.com/opencv/opencv_extra/blob/5.x/testdata/highgui/drawing/text_test.png), however some further polishing is needed:
  * update stb-based rendered from the fresh upstream version (or use freetype renderer, as in the original PR)
  * use more robust method to handle right-to-left scripts and/or scripts with ligatures, like Arabic, Devangari etc. The most natural method is to use Harfbuzz (https://github.com/harfbuzz/harfbuzz). Support for Harfbuzz was included into the original patch, but then was excluded to reduce footprint.
  * fonts (currently free Rubik font is used) likely need to be updated to support more characters out of the box, e.g. DroidSans-Fallback.

### Computational Geometry

This functionality is not as widely used as `cv::resize()`, but it's still fundamental and important. Let's keep it and improve.

* We probably need more algorithms to extract vector contours, e.g. from the output of semantic segmentation or instance segmentation deep models or Canny edge detector. In some specialized packages there is notion of 'vector tracing' that we probably need to implement.

* Once a contour or a 2D point cloud is obtained, the current functionality, like convexHull or fitEllipse, is fine, but we need more basic operations to extract a part of contour or a subset of point cloud.

* Maybe Delaunay triangulation should be moved to opencv_contrib.

* As described in #25001, it's suggested to move all Hough transform-related functionality into opencv_contrib. Instead, we need efficient compact deep models to detect geometric primitives (lines, circles, ellipses). Of course, it means dependency of DNN module and so this new functionality shall be put into another module, e.g. objdetect.

**Comment by vpisarev** [comment]
Some work on adding FP16 support and on acceleration of imgproc is being done:
#25984, #26271

**Comment by kinchungwong** [comment]
### Image warping

Recently I have started experimenting with alternative designs for the general remap function. This is my attempt to solve the long-standing #7544 issue (32k limit). I'd also like to take a jab at @nh2 's [snarky comment](https://github.com/opencv/opencv/issues/7544#issuecomment-2424598062). Could it be the case that we can actually use a lambda?

I only started December, so I totally missed the boat for OpenCV 5.0. Congratulations to the team, and sorry for not being able to contribute earlier. I also wasn't aware of this discussion thread when I started; my work might have duplicated with something that's already done and shipped.

Currently there is nothing much to show, but here's the repository. https://github.com/kinchungwong/opencv_remap_toy  (Don't try to build it; right now it only builds on my home computer.) 

I'm now prioritizing adding design documents over actual coding. As a result, the interesting stuff are going to be in the ```doc``` folder. If anyone brings up copyright issue, these design documents will become the artifact that enable others to do clean development.

During my work I realized some of these stuff could result in useful HAL kernels that can be contributed back to OpenCV. That's why I want to raise awareness now and ask for advice.

The code is currently AVX-512 only, because part of my current work is to explore the suitability and performance of AVX2 and AVX-512 gather instructions on both Intel and AMD with respect to image undistortion workload. There's no benchmark yet. Right now I'm working toward completing an end-to-end 8UC1 bicubic and radial undistortion, and then start analyzing the machine code and clock cycles.

The long plan is to bring it to other SIMD architectures. Universal intrinsics will be used where possible; if not, dispatch will be used.

### Disclaimer and some historical backgrounds for others.

Disclaimer: I am the original author of the tiling remap gist. The idea was simple, tiling. 

I *do not* recommend anyone read the code (if they haven't). For those who need to, here it is: https://gist.github.com/kinchungwong/141bfa2d996cb5ae8f42

Unfortunately my former employer never released the copyright to that code; I am now coding everything straight up from scratch. I anticipate that, later in the process, I could benefit from other participants, who can either enhance the clean-roomliness of new code contributions, or become neutral party code similarity reviewers to clear the new contributions of potential concerns.)

Personally I don't think there's any patent issue with just the idea of tiling, so feel free for anyone to implement with their own code.

**Comment by smissan** [comment]
The remap limitations (to me) are the biggest problem in the current opencv. I agree that creating a "mapping" Mat is wasteful in most cases and lambda sounds great in principle. 

Whatever the new solution is going to be proposed it should also support F32 Mats and borders including border reflect. 

I'm not an expert in SIMD but would like to keep my eye on this development, so glad that the progress is finally being made.

**Comment by crackwitz** [comment]
The point of remap is to work with such lookup tables. Some mappings cannot be represented other than with lookup tables.

if you need something that is different from remap(), then it's not remap() that you need.

**Comment by smissan** [comment]
exactly, I want image wraps that are not using lookup tables and have no size limits and extra allocations. But because these methods currently call remap, we are stuck with remap limitations.

**Comment by nh2** [comment]
> if you need something that is different from remap(), such as affine or perspective warps, then it's not remap() that you need, but those specific things.

@crackwitz I do not fully understand the point. There seem to be lots of cases where a lookup table isn't needed and a lambda would suffice, but which aren't specific enough to become their own OpenCV functions. For example, any "pure" coordinate-transforming remapping function where you need to say where in the old image an XY pixel from the new image is to be found.

**Comment by crackwitz** [comment]
> a lambda

if you mean a "numerical kernel" by that, calculating something for each array element, then that already exists: OpenCV has a parallel `for` construct. IDK if that can turn lambdas into OpenCL code.

**Comment by nh2** [comment]
> That also expects you to implement the sampling itself

Right, ideally remap would do that for me, making he handle only the coordinate transform.

---

I explored a middle path for reducing memory usage, in between the full `map` lookup table and the per-pixel lambda:

Calling `remap()` repeatedly, incrementally on blocks of rows ("tiling" on the row-axis).

It works quite well and preserves the existing acceleration approach, you make the blocks large enough:

https://github.com/nh2/opencv-incremental-remap-benchmark

It also looks like this approach produces identical images to full `remap()`.

Maybe the docs of `remap()` should be extended to:

* _guarantee_ that tiled remaps have identical results
* suggest to the user to use this technique to remove memory usage

**Comment by vpisarev** [comment]
TBD: need to add 'TODO' checklist in the first message of the ticket.
